server/src/test/java/org/elasticsearch/action/ActionListenerTests.java

```diff
@@ -370,6 +370,40 @@ public void onFailure(Exception e) {
         assertThat(exReference.get(), instanceOf(IllegalArgumentException.class));
     }
 
+    public void testAssertAtLeastOnceWillLogAssertionErrorWhenNotResolved() throws Exception {
+        assumeTrue("assertAtLeastOnce will be a no-op when assertions are disabled", Assertions.ENABLED);
+        final AtomicReference<ActionListener<Object>> listenerRef = new AtomicReference<>(
+            ActionListener.assertAtLeastOnce(ActionListener.running(() -> {
+                // Do nothing, but don't use ActionListener.noop() as it'll never be garbage collected
+            }))
```

Comment:
This is perhaps a shortcoming, if the listener is something that is retained in some other way (such as the listener returned by `noop()`) the detection won't work because it'll never become unreachable. I'm not sure how common this is though. We could insert another pass-through delegate in the chain that isn't referenced from anywhere to make it more robust, but that would allocated an additional object for what might be a rare edge-case.